firewood/src/storage/linear/mod.rs

```diff
@@ -96,3 +96,99 @@ impl<S: ReadLinearStore> ReadLinearStore for LinearStore<S> {
+                self.bytes.resize(offset + object.len(), 0);
+            }
+            self.bytes[offset..offset + object.len()].copy_from_slice(object);
+            Ok(object.len())
+        }
+    }
+
+    impl ReadLinearStore for InMemReadWriteLinearStore {
+        fn stream_from(&self, addr: u64) -> Result<impl Read, std::io::Error> {
+            Ok(&self.bytes[addr as usize..])
+        }
+
+        fn size(&self) -> Result<u64, std::io::Error> {
+            Ok(self.bytes.len() as u64)
+        }
+    }
+
+    #[test]
+    fn test_in_mem_write_linear_store() {
+        let mut store = InMemReadWriteLinearStore { bytes: vec![] };
+        assert_eq!(store.size().unwrap(), 0);
+
+        // Write to an empty store
+        store.write(0, &[1, 2, 3]).unwrap();
+        assert_eq!(store.bytes, vec![1, 2, 3]);
+        assert_eq!(store.size().unwrap(), 3);
+
+        // Write at the end of the store
+        store.write(3, &[4, 5, 6]).unwrap();
+        assert_eq!(store.bytes, vec![1, 2, 3, 4, 5, 6]);
+        assert_eq!(store.size().unwrap(), 6);
+
+        // Overwrite the beginning of the store
+        store.write(0, &[7, 8]).unwrap();
+        assert_eq!(store.bytes, vec![7, 8, 3, 4, 5, 6]);
+        assert_eq!(store.size().unwrap(), 6);
+
+        // Overwrite the middle of the store
+        store.write(1, &[9, 10]).unwrap();
+        assert_eq!(store.bytes, vec![7, 9, 10, 4, 5, 6]);
+        assert_eq!(store.size().unwrap(), 6);
+
+        // Write past the end of the store
+        store.write(7, &[11, 12]).unwrap();
+        assert_eq!(store.bytes, vec![7, 9, 10, 4, 5, 6, 0, 11, 12]);
+        assert_eq!(store.size().unwrap(), 9);
+
+        // Read from the start of the store
+        let mut reader = store.stream_from(0).unwrap();
+        let mut read_bytes = vec![];
+        let num_read_bytes = reader.read_to_end(&mut read_bytes).unwrap();
+        assert_eq!(read_bytes, vec![7, 9, 10, 4, 5, 6, 0, 11, 12]);
+        assert_eq!(num_read_bytes, 9);
+
+        // Read from the middle of the store
+        let mut reader = store.stream_from(3).unwrap();
+        let mut read_bytes = vec![];
+        let num_read_bytes = reader.read_to_end(&mut read_bytes).unwrap();
+        assert_eq!(read_bytes, vec![4, 5, 6, 0, 11, 12]);
+        assert_eq!(num_read_bytes, 6);
+
+        // Read from the end of the store
+        let mut reader = store.stream_from(store.size().unwrap() - 1).unwrap();
+        let mut read_bytes = vec![];
+        let num_read_bytes = reader.read_to_end(&mut read_bytes).unwrap();
+        assert_eq!(read_bytes, vec![12]);
+        assert_eq!(num_read_bytes, 1);
+
+        // Read from past the end of the store
+        let mut reader = store.stream_from(store.size().unwrap()).unwrap();
+        let mut read_bytes = vec![];
+        let num_read_bytes = reader.read_to_end(&mut read_bytes).unwrap();
+        assert_eq!(read_bytes, vec![]);
+        assert_eq!(num_read_bytes, 0);
+    }
```

Comment:
These should all be separate tests. If the first test fails, you'll have no idea how many (if any) the other tests fail and will have to fix each test one at a time instead of all the different tests that fail at once.